### PR TITLE
GraphQL - Add Posts and Comments to User

### PR DESCRIPTION
**Pull request by toyhammered**
# What

Add ability to query post/comments and pinned post from a user.

# Why

# Checklist

- [ ] All files pass Rubocop
- [ ] Any complex logic is commented
- [ ] Any new systems have thorough documentation
- [ ] Any user-facing changes are behind a feature flag (or: explain why they can't be)

- [ ] All the tests pass
- [ ] Tests have been added to cover the new code

### CODE REVIEW

File: app/graphql/types/comment.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  field :id, ID, null: false
+
+  field :post, Types::Post,
+    null: false,
+    description: 'The post that this comment is attached to.'
+
+  field :profile, Types::Profile,
+    null: false,
+    description: 'The owner of this post',
+    method: :user
```
Comment (NuckChorris) [inline app/graphql/types/comment.rb]: You don't necessarily want the field to be named just `profile`

You might instead consider something descriptive like `creator`

File: app/graphql/types/comment.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    null: false,
+    description: 'Html formatted content.'
+
+  field :parent, Types::Comment,
+    null: true,
+    description: 'The parent comment if this comment was a reply to another.'
+
+  field :likes, Types::Profile.connection_type,
+    null: false,
+    description: 'Users who liked this comment.'
```
Comment (NuckChorris) [inline app/graphql/types/comment.rb:29]: I think this is reasonable to hide the Like object, if we need more advanced shit like multiple types of "likes" we can add it later.  Easier to add than remove!

File: app/graphql/types/comment.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+  field :replies, Types::Comment.connection_type,
+    null: false,
+    description: 'All replies to a specific comment.'
+
+  def likes
+    AssociationLoader.for(object.class, :likes).load(object).then do |likes|
+      likes.map(&:user)
```
Comment (NuckChorris) [inline app/graphql/types/comment.rb]: This is an N+1 problem, you need to use a loader to get the User

File: app/graphql/types/post.rb [hunk 4]
```diff
@@ -0,0 +1,63 @@
+class Types::Post < Types::BaseObject
+  description 'A post that is visible to your followers and globally in the news-feed.'
+
+  field :id, ID, null: false
+
+  field :profile, Types::Profile,
```
Comment (NuckChorris) [inline app/graphql/types/post.rb]: As above, I think you should be more descriptive here

File: app/graphql/types/post.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+  field :media, Types::Media,
+    null: true,
+    description: 'The media tagged in this post.'
+
+  field :spoiler, Boolean,
+    null: false,
+    description: 'If this post spoils the tagged media.'
+
+  field :nsfw, Boolean,
```
Comment (NuckChorris) [inline app/graphql/types/post.rb]: I think graphQL generally puts `is` prefixes on these fields like `isSpoiler` or `isNsfw`

File: app/graphql/types/post.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+  def likes
+    AssociationLoader.for(object.class, :post_likes).load(object).then do |likes|
+      likes.map(&:user)
+    end
+  end
+
+  def follows
+    AssociationLoader.for(object.class, :post_follows).load(object).then do |follows|
+      follows.map(&:user)
```
Comment (NuckChorris) [inline app/graphql/types/post.rb]: Here and a few lines up where you do `likes.map` are both N+1 too!  AssociationLoader only loads `post->likes/post_follows`, you need to layer another AssociationLoader to get `post->likes->user`

**Review by NuckChorris — CHANGES_REQUESTED** [review]

File: app/graphql/types/comment.rb (hunk 1 above)
Comment (toyhammered) [reply app/graphql/types/comment.rb]: Ok, I'll go through all the places I use a profile and change them to make more sense per each context.

Comment (toyhammered) [reply app/graphql/types/comment.rb]: Debating between `creator` and `owner` 🤔

Comment (wopian) [reply app/graphql/types/comment.rb]: `author`?

**Review by NuckChorris — APPROVED** [review]

**Review by wopian — APPROVED** [review]

**Pull request merged**